service/history/visibility_queue_task_executor.go

```diff
@@ -305,12 +308,26 @@ func (t *visibilityQueueTaskExecutor) processCloseExecution(
 	// Therefore, ctx timeout might be already expired
 	// and parentCtx (which doesn't have timeout) must be used everywhere bellow.
 
-	if t.enableCloseWorkflowCleanup(namespaceEntry.Name().String()) {
+	if t.needRunCleanUp(requestBase) {
 		return t.cleanupExecutionInfo(parentCtx, task)
 	}
 	return nil
 }
 
+func (t *visibilityQueueTaskExecutor) needRunCleanUp(
+	request *manager.VisibilityRequestBase,
+) bool {
+	if !t.enableCloseWorkflowCleanup(request.Namespace.String()) {
+		return false
+	}
+	// If there are no memo nor search attributes, then no clean up is necessary.
+	if len(request.Memo.GetFields()) == 0 && len(request.SearchAttributes.GetIndexedFields()) == 0 {
+		return false
+	}
+	minSize := t.relocateAttributesMinBlobSize(request.Namespace.String())
+	return request.Memo.Size() >= minSize || request.SearchAttributes.Size() >= minSize
```

Comment:
shall we check if memo_size + sa_size > minSize ?

Reply:
I'm ok either way.